### ISSUE TITLE
Sealing off non-SC library calls -- separate ST versus IO with modified PrimMonad class

### ISSUE THREAD

**Issue by rrnewton** [iu-parfunc/sc-haskell#1]
Today @vollmerm and I poked at the `primitive` package a little bit.

We found that there is currently no real operational distinction between `PrimMonad`'s `IO` and `ST` instances when it comes to issuing a write---they issue the same `writeMut*#` primops.

The proposal in this issue is to change `PrimMonad` so that there is are separate entrypoints for (at least) read and write primitive operations.  That way on x86 the IO write operation can include a fence, while the ST one will not.

**Comment by rrnewton** [comment]
Well, actually, it's really not a "read vs write" thing.  Rather the current `primitive` method would be split into a variety of categories related to the SC-Haskell goal. Basically it would be these I think:
- Needs a fence on TSO
- Needs a fence on Power/ARM but not TSO
- Never needs a fence.

Then the `ST` instance of `PrimMonad` omits all fences of course.

**Comment by vollmerm** [comment]
I started playing with making an alternate module for `IORef`, but changing `PrimMonad` might be more difficult, since the benchmark programs will use it indirectly (via `Vector` or whatever). We would maybe have to provide an alternate version of the primitive package and compile our benchmark programs using it.

**Comment by rrnewton** [comment]
Yeah, but it would be awful nice if we actually didn't have to change `vector` (much), and only had to change the `primitive` package that it depends on....  alas, there will be some change, replacing the `primitive` function call with something more specific.

**Comment by vollmerm** [comment]
I agree it would be great if we could just replace `primitive` and have everything else fall out.

I looked into using GHC pragmas to specialize the `PrimMonad` instances, but it looks like the functionality is limited. You can't specify a new function to use when specializing a generic function, though apparently that used to be possible. You _could_ achieve the same thing using rewrite rules, I think, so that might get us what we want.

**Comment by madanMus** [comment]
Hi Ryan, Mike,
 Sorry, emails from GitHub are going to an email address that I do not check often. So the delay. Regarding the issue Ryan brought up the different categories of operations one needs, the Java Cookbook by Doug Lea (http://gee.cs.oswego.edu/dl/jmm/cookbook.html) is a good standard to follow. We have to determine which of the loads and stores in Haskell have the volatile semantics and the cookbook will tell us how to convert these operations into different architectures.

**Comment by rrnewton** [comment]
Thanks!  Useful reference.

**Comment by rrnewton** [comment]
Heh, in fact I should have read this before because I see now [that it's linked from the comments](https://github.com/ghc/ghc/blob/343349df1f19f21899818d647bf570e489d0cf8f/includes/stg/SMP.h#L89) in GHC's barrier code already.

@vollmerm, it seems reasonable to bifurcate the `primitive` method into one version for each barrier type: LoadStore, LoadLoad, StoreStore, StoreLoad.  The fifth one can just be the current `primitive` which implies no extra barrier.  Then we just need to audit each occurrence of `primitive` in, e.g., `vector` to see if it should be `primitiveLSB`, `primitiveLLB`, `primitiveSSB`, or `primitiveSLB` instead.  (I'm fine with long names too, if you like: `primitiveLoadStoreBarrier`.)

In the GHC RTS code linked above, and then [exposed by atomic-primops](https://hackage.haskell.org/package/atomic-primops-0.8.0.3/docs/Data-Atomics.html#g:6), we've got portable versions (i.e. internally ifdef'd on architecture) of three of those four barriers.  We're missing LoadStore for some reason.

**Comment by rrnewton** [comment]
@vollmerm: one more thing, I think it's fine to also expose just the barriers by themselves in the class:

``` Haskell
class Monad m => PrimMonad m where
  type PrimState m
  primitive :: (State# (PrimState m) -> (# State# (PrimState m), a #)) -> m a
  loadStoreBarrier :: m ()
  loadLoadBarrier :: m () 
  storeStoreBarrier :: m ()
  storeLoadBarrier :: m () 
```

You can do it either (1) this way, or (2) by making those variants of `primitive` discussed above (i.e. with the same type as primitive).  Those would perform the designated barrier _before_ the given operation.  The barriers by themselves are fine...  it will just mean using the monad's bind to compose the barrier in front of the primitive action manually: `do storeLoadBarrier; primitive foo`.  It's just one more bind that the compiler must inline and optimize away.

There's no difference in expressiveness, because:

``` Haskell
storeLoadBarrier == primitiveSLB (\s -> (# s, () #))
```

**Comment by vollmerm** [comment]
I [forked a copy of `primitive`](https://github.com/vollmerm/primitive) and started experimenting with adding stuff to `PrimMonad`.

BTW, it looks like the ["JSR-133 Cookbook for Compiler Writers"](http://gee.cs.oswego.edu/dl/jmm/cookbook.html) is also referenced in [GHC's SMP.h](https://github.com/ghc/ghc/blob/master/includes/stg/SMP.h), where the barrier functions are described.

**Comment by vollmerm** [comment]
I'm trying to page this stuff back. @rrnewton what did we decide to do with `primitive` in the meeting last Monday? Somehow we're going to set it up so that a library like `Vector` that calls `primitive` and then does stuff with mutable arrays will be safe, right?

I guess we are still stalled on figuring out how to add a primop for a fence and determining where/how the compiler reorders stuff in CMM.

**Comment by rrnewton** [comment]
@vollmerm, yes you are testing a proposed permanent modification for `primitive`, so along with running your rebuilt GHC with a different wired-in `base` package, you should arrange to run your fork of the `primitive` package as well.  You'll probably have to do a dance to get software building, such as a stackage, while substituting your GHC and your primitive package.  Stack should work pretty well with this however, as you just point at a local copy of `primitive` and it will override the one in the stackage LTS snapshot.

You're not stalled, however.  Go ahead and have these primops call the non-inlined barriers from `atomic-primops`.  Try to get us to a point where you have a concrete performance problem, and then we can block on fixing it (i.e. #4) if we need to.

Tentatively assume for the time being that CMM optimizations do nothing wrong, and we'll figure out of that assumption is wrong and fix it when we can (#5).  Let's do the easier things first and get some initial results, rather than getting stalled on the hardest parts.

**Comment by vollmerm** [comment]
Getting back into the swing of things here after the semester end...

Right now the changes are live in my GHC fork to just handle `IORef` and `IOArray` specially, and this is enough to run `nofib`. We can look at the `nofib-analyse` results on Monday, maybe, but there does seem to be a slowdown.

**Comment by vollmerm** [comment]
@rrnewton: I exported `storeLoadBarrier` in `GHC.Barrier` and my fork of `primitive` now places a barrier after writes to `Array`, `ByteArray`, and `MutVar` when they are in `IO`.

My `primitive` fork is linked as a submodule in the `iu-parfunc/ghc` repo, and I think the docker image is building, so it should be ready to go for testing/benchmarking.
